### PR TITLE
Feature/swap formula

### PR DESCRIPTION
**Pull request by joneskm**

### CODE REVIEW

**Review by AustinoBombino — APPROVED** [review]
Looks good, nice job

File: x/clp/keeper/msg_server.go [hunk 1]
```diff
@@ -998,3 +1001,30 @@ func (k msgServer) ModifyLiquidityProtectionRates(goCtx context.Context, msg *ty
...
+
+func (k msgServer) UpdateSwapFeeRate(goCtx context.Context, msg *types.MsgUpdateSwapFeeRateRequest) (*types.MsgUpdateSwapFeeRateResponse, error) {
+	response := &types.MsgUpdateSwapFeeRateResponse{}
+
+	// defensive programming
+	if msg == nil {
+		return response, errors.Errorf("msg was nil")
+	}
+
+	if err := msg.ValidateBasic(); err != nil {
```
Comment (Jedi2002) [inline x/clp/keeper/msg_server.go:1013]: Just curious, is there any way you can think of, through which someone can reach this point of the code, without executing validate basic once already

File: x/clp/keeper/pureCalculation_test.go [hunk 2]
```diff
@@ -97,13 +98,42 @@ func TestKeeper_RatToDec(t *testing.T) {
...
 			require.Equal(t, tc.expected, y)
 		})
 	}
 }
 
+// If the cosmos SDK reduces the value of maxDecBitLen below 315 then (without
+// further changes) RatToDec will be able to return an sdk.Dec with 316 bits.
+// Such sdk.Decs would cause panics and errors when unmarshaling. This test
+// will fail if the cosmos SDK reduces the value of maxDecBitLen < 315.
+func TestKeeper_RatToDec_Overflow(t *testing.T) {
```
Comment (Jedi2002) [inline x/clp/keeper/pureCalculation_test.go:118]: Nice !

File: x/tokenregistry/types/msgs.go [hunk 3]
```diff
@@ -38,8 +38,8 @@ func (m *MsgRegister) ValidateBasic() error {
 	if err != nil {
 		return sdkerrors.Wrap(err, "invalid from address")
 	}
-	if m.Entry.Decimals <= 0 {
-		return errors.New("Decimals cannot be zero")
+	if m.Entry.Decimals < 0 {
```
Comment (Jedi2002) [inline x/tokenregistry/types/msgs.go:41]: Are there any tokens that have 0 decimals? 
Might be a good idea to keep this in even if the code can handle 0 decimal tokens.

**Pull request merged**